### ISSUE TITLE
Handling of Z flag in defaultMsgAcceptFunc

### ISSUE THREAD

**Issue by chantra** [miekg/dns#975]
Following the thread [[dns-operations] The last remaining DNS header flag](https://lists.dns-oarc.net/pipermail/dns-operations/2019-May/018724.html) it seems rather than [rejecting the Message](https://github.com/miekg/dns/blob/5c2ec9f7e458912d656335743784d0195a5a423e/acceptfunc.go#L38), the Z bits should be reset to 0 and the message processed just like if nothing was wrongly set.
While `RFC1035 4.1.1. Header section format` does not seem to say much other than:
```
Z               Reserved for future use.  Must be zero in all queries
                and responses.
```
and in this case it is non-zero in the query, it would make sense for "future use" that whatever they may be set to, the response would just set them to 0 as a signal that it does not support any specific use of (currently) Z flags.

https://ednscomp.isc.org currently return the following warning when the response echo the bits:

```
Plain DNS with last reserved header bit set (zflag)
dig +norec +noad +noedns +zflag soa zone @server
expect: SOA
expect: NOERROR
expect: Z bit to be clear in response
See RFC1035, 4.1.1. Header section format
```

A simple patch that would make this work would be:

```patch
@@ -25,6 +25,7 @@
 )
 
 func defaultMsgAcceptFunc(dh Header) MsgAcceptAction {
+	dh.Bits = dh.Bits & ^uint16(_Z)
 	if isResponse := dh.Bits&_QR != 0; isResponse {
 		return MsgIgnore
 	}
@@ -35,9 +36,6 @@
 		return MsgReject
 	}
 
-	if isZero := dh.Bits&_Z != 0; isZero {
-		return MsgReject
-	}
 	if dh.Qdcount != 1 {
 		return MsgReject
 	}
```
but this would make the flag cleared before the request is actually handled, which may not be the best.
Another approach could be to clear the bits when setting the response (which would make more sense IMHO), like in the helper functions in https://github.com/miekg/dns/blob/d051b464e9f024b65ce9674dece9d3e905b31828/defaults.go

Throughts? cc @yarikk

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[miekg/dns] Handling of Z flag in d..." ]
>Following the thread [[dns-operations] The last remaining DNS header flag](https://lists.dns-oarc.net/pipermail/dns-operations/2019-May/018724.html) it seems rather than [rejecting the Message](https://github.com/miekg/dns/blob/5c2ec9f7e458912d656335743784d0195a5a423e/acceptfunc.go#L38), the Z bits should be reset to 0 and the message processed just like if nothing was wrongly set.

Seems like an easy change to make. Would be nice if Mark would write an RFC for 
this though

**Comment by chantra** [comment]
I posted on the mailing list https://lists.dns-oarc.net/pipermail/dns-operations/2019-May/018773.html

to which Mark replied:
> Formerr w/o the zflag is also acceptable.  STD13 allows for both behaviours.
That said most servers return NOERROR.

I think it may just be fine to let the query go through and not check the flags, yet reset them when writing with the helpers.

This leave the implementer of ServeDNS a chance to reject if flags are set in query, yet provide some helpers that will do they best to set the response with Z flag reset.

On another note, https://github.com/miekg/dns/blob/5c2ec9f7e458912d656335743784d0195a5a423e/acceptfunc.go#L34 should rather return a NotImplemented response rather than a reject. I was checking at other DNS compliance testing tests and stuff like:
```
dig +noedns +noad +norec +header-only +opcode=15
```
should return NotImpl with opcode left to 15. Current implementation would return `opcode: QUERY, status: FORMERR`

**Comment by tmthrgd** [comment]
I don’t think zeroing can happen in `defaultMsgAcceptFunc` as it isn’t given a pointer to `Header`.

Is the Z bit used by any existing software? Perhaps a solution is to deprecate the Zero field in `MsgHdr` and always leave it unset in `setHdr` and ignore it in `packBufferWithCompressionMap`.

If we don’t want to go down that root, zeroing it in the helper `*Msg` functions seems the only way. It’s passed to the handler and can be set in the written message. It would seem very odd to me to only be able to read it or only be able to write it.

`MsgAcceptAction` could be extended to include a not implemented return if that’s worthwhile.

**Comment by miekg** [comment]
when can remove the check and just accept it.

As @tmthrgd says, zeroing is hard to do and should probably not happing in this library. Esp when folks then want to set it, so this will then also require an escape hatch.

**Comment by yarikk** [comment]
Please see PR #976: it accepts packets ignoring the value of Z flag, and explicitly clears it out in responses generated by the library, otherwise retaining the library transparent on the regards in a forward-compatible way (so it seems).

**Comment by chantra** [comment]
Actually, `SetReply` https://github.com/miekg/dns/blob/d051b464e9f024b65ce9674dece9d3e905b31828/defaults.go#L15 does not copy the flag over, so that's good as it means that the user of the library will be responsible of setting those flags if it needs to, and by default they will be cleared.
The case of invalid request on the other hand directly reuse the request message and set the RCODE to FormErr. @yarikk 's patch would clear the flag in that case, which is probably a good default.

**Comment by miekg** [comment]
closed via 15450720
